### PR TITLE
Pseudo-huber loss metric added

### LINKED ISSUE #5479: Native Pseudo-Huber loss support
Would it be possible to support Pseudo-Huber-loss (https://en.wikipedia.org/wiki/Huber_loss#Pseudo-Huber_loss_function) natively?

I implemented it as a custom losss function (I use the Python SKLearn API)
``` python
def huber_approx_obj(y_true, y_pred):
    z = y_pred - y_true
    delta = 1
    scale = 1 + (z/delta)**2
    scale_sqrt = np.sqrt(scale)
    grad = z/scale_sqrt
    hess = 1/(scale*scale_sqrt)
    return grad, hess
```
but the feature importance plots don't support custom loss functions (and it slows the learning process in comparison to 'reg:squarederror'). 

The basic problem is the need for a robust regession objective; MSE can be sensitive to outliers in application.

### PR DESCRIPTION
**Pull request by LionOrCatThatIsTheQuestion**
Fixes #5479 .

@trivialfis

### CODE REVIEW

**Comment by trivialfis** [comment]
You can use `XGBoostParameter` for defining the delta, see `objective/multiclass_obj.cu::SoftmaxMultiClassParam` for an example.

**Comment by LionOrCatThatIsTheQuestion** [comment]
Added the registration in `regression_obj.cu` similar to squared errror. Not sure if that's it?

**Comment by LionOrCatThatIsTheQuestion** [comment]
The parameter is way over my head of understanding C/ CUDA code - I can see how you register a parameter in `regression_obj.cu`, but wouldn't there have to be a class similar to `PoissonRegression` in the same file?

**Comment by trivialfis** [comment]
@LionOrCatThatIsTheQuestion Don't worry.  I'm happy to merge the PR provided you can add tests in `/tests/cpp/objective/test_regression_obj.cc` and similar place for metric.  Also I need to go through the derivation later today.

But if you want to give it another go for doing the parameter.  It's a `factory` method, which is also available in Python and is heavily used by libraries like pytorch and dask.  I found a very long explanation for Python: https://realpython.com/factory-method-python/  But you can simply search for `Python factor` and there are lots of introduction.  It works similarly in C++, just the macros hiding many details so the code seems a  little bit ridiculous.

**Comment by LionOrCatThatIsTheQuestion** [comment]
The tests are added, I computed the reference values with Wolfram Alpha - really hope there aren't any more typos, I'll check back tomorrow :).

**Comment by LionOrCatThatIsTheQuestion** [comment]
To answer to the parameter part, I took a short look at the factory methods and if I have time left at the end of my current project, I'll come back to your explanantion and try again with a new PR :).

**Comment by trivialfis** [comment]
@LionOrCatThatIsTheQuestion  Just checked the derivation.  Good work!  Now one last piece, please add documentation on `doc/parameter.rst` `objective` section and `eval_metric` section.

**Comment by LionOrCatThatIsTheQuestion** [comment]
Documentation done as well, thank you very much for guiding me through the code :).

**Comment by trivialfis** [comment]
@LionOrCatThatIsTheQuestion Thanks!

**Comment by cristianogoncalves** [comment]
@LionOrCatThatIsTheQuestion  Excellent feature implemented to xgboost. Is it possible to select \delta paramater or it´s 1 ?   =)

**Comment by trivialfis** [comment]
It's just 1 for now, but should be easily extended.  I'm not sure how urgent it's to get different slope for this function.

File: src/objective/regression_loss.h [hunk 1]
```diff
@@ -98,6 +98,37 @@ struct LogisticRegression {
...
+struct PseudoHuberError {
+  XGBOOST_DEVICE static bst_float PredTransform(bst_float x) {
+    return x;
+  }
+  XGBOOST_DEVICE static bool CheckLabel(bst_float label) {
+    return true;
+  }
+  XGBOOST_DEVICE static bst_float FirstOrderGradient(bst_float predt, bst_float label) {
+    const float z = predt - label;
+    const float scale_sqrt = std::sqrt(1 + std::pow(z, 2));
```
Comment (QuantHao) [inline src/objective/regression_loss.h:110]: @

**Comment by QuantHao** [comment]
> It's just 1 for now, but should be easily extended. I'm not sure how urgent it's to get different slope for this function.

Hi, @trivialfis thanks for support this feature! Based on sklearn's suggestion, it seems that delta is better at 1.35 to achieve 95% statistical efficiency.
https://scikit-learn.org/stable/modules/linear_model.html#huber-regression

**Comment by trivialfis** [comment]
@LionOrCatThatIsTheQuestion Would you like me to carry it out?  :-)

**Comment by QuantHao** [comment]
@trivialfis @LionOrCatThatIsTheQuestion After reviewing the code, it seems to me that implemented delta of 1 actually depends on variance of dependent variable y, which is similar to SGDRegressor with huber loss. By comparision, HuberRegressor in sklearn says that its delta is scaling invariant. Reference: https://scikit-learn.org/stable/modules/linear_model.html#notes

Therefore, users should standardize y to ensure variance as 1. Is my understanding correct? Thanks. :)

**Comment by QuantHao** [comment]
@LionOrCatThatIsTheQuestion @trivialfis Hi, I review the XGboost GPU supported metrics documents and it seems that this huber loss is current not compatible with GPU training. Am I correct? https://xgboost.readthedocs.io/en/latest/gpu/index.html#objective-functions

**Comment by hcho3** [comment]
@QuantHao The list should be updated, since the pseudo-Huber loss is supported by the GPU algorithm.

**Comment by hcho3** [comment]
I filed a pull request to update the list: https://github.com/dmlc/xgboost/pull/5909

**Comment by QuantHao** [comment]
> @QuantHao The list should be updated, since the pseudo-Huber loss is supported by the GPU algorithm.

Thanks so much for your confirmation.

**Comment by QuantHao** [comment]
@hcho3 Also, it would be great to update "Supported parameters" section for GPU support documentation. It seems that some parameters are missing such as colsample_bynode, alpha, lambda, etc. Thanks.
https://xgboost.readthedocs.io/en/latest/gpu/index.html#supported-parameters
